### PR TITLE
support endless recorder for timeline

### PR DESCRIPTION
**Pull request by yuanhuihui**
Timeline use TimelineEventRingRecorder by default.
If we set trace-systrace, using TimelineEventSystraceRecorder.
If set endless-trace-buffer, we need support  using TimelineEventEndlessRecorder.

in file third_party/dart/runtime/vm/timeline.cc
```
static TimelineEventRecorder* CreateTimelineRecorder() {
  const bool use_endless_recorder =
      (FLAG_timeline_dir != NULL) || FLAG_timing || FLAG_complete_timeline;

  const bool use_startup_recorder = FLAG_startup_timeline;
  const bool use_systrace_recorder = FLAG_systrace_timeline;
  const char* flag = FLAG_timeline_recorder;

  if (use_systrace_recorder || (flag != NULL)) {
    if (use_systrace_recorder || (strcmp("systrace", flag) == 0)) {
#if defined(HOST_OS_LINUX) || defined(HOST_OS_ANDROID)
      return new TimelineEventSystraceRecorder();
#elif defined(HOST_OS_FUCHSIA)
      return new TimelineEventFuchsiaRecorder();
#else
      OS::PrintErr(
          "Warning: The systrace timeline recorder is equivalent to the"
          "ring recorder on this platform.");
      return new TimelineEventRingRecorder();
#endif
    }
  }

  if (use_endless_recorder || (flag != NULL)) {
    if (use_endless_recorder || (strcmp("endless", flag) == 0)) {
      return new TimelineEventEndlessRecorder();
    }
  }

  if (use_startup_recorder || (flag != NULL)) {
    if (use_startup_recorder || (strcmp("startup", flag) == 0)) {
      return new TimelineEventStartupRecorder();
    }
  }
  return new TimelineEventRingRecorder();
}
```

### CODE REVIEW

**Comment by zanderso** [comment]
Is this flag in the engine yet? I'll mark this as "work in progress", but feel free to unmark and add reviewers when it is ready. Thanks!

**Comment by yuanhuihui** [comment]
> Is this flag in the engine yet? I'll mark this as "work in progress", but feel free to unmark and add reviewers when it is ready. Thanks!

Yes, this flag is already in the engine, it works now.  how can I unmark "work in progress" ?

switches.cc

**Comment by yuanhuihui** [comment]
https://github.com/flutter/engine/pull/14568

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -103,6 +103,13 @@ class RunCommand extends RunCommandBase {
         help: 'Enable tracing of Skia code. This is useful when debugging '
               'the GPU thread. By default, Flutter will not log skia code.',
       )
+      ..addFlag('endless-trace-buffer',
+        negatable: false,
+        help: 'Enable tracing to the endless tracer. This is usefull when '
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/commands/run.dart]: nit: Change `usefull` to `useful`.

**Comment by cbracken** [comment]
@yjbanov I'm assuming that we'll need to update our devicelab tests to set this new flag where necessary?

**Comment by yjbanov** [comment]
> @yjbanov I'm assuming that we'll need to update our devicelab tests to set this new flag where necessary?

I'd say very likely.

**Comment by yuanhuihui** [comment]
@googlebot I fixed it

**Comment by yuanhuihui** [comment]
My accout have sign CLA success. This third commit was created on GitHub.com and signed with another mail(qq email) accidentally, but this two email correspond to one github account. I don't know how to sign CLA for another email, what should I do?  or I resubmit a new PR?

**Comment by yuanhuihui** [comment]
relations to https://github.com/flutter/flutter/issues/47771

**Comment by yuanhuihui** [comment]
> > @yjbanov I'm assuming that we'll need to update our devicelab tests to set this new flag where necessary?
> 
> I'd say very likely.

yes. but this PRThis pr just exposes one more parameter, no matter the engine modifies the startup logic

**Comment by yuanhuihui** [comment]
@zanderso @cbracken @yjbanov could you review it?

**Comment by zanderso** [comment]
@cbracken @yjbanov Not sure I understand your comments. Do you think the device lab tests should be updated as part of this PR or in the future?

**Comment by jonahwilliams** [comment]
What is the status of this PR, are we waiting for an engine roll?

**Comment by jonahwilliams** [comment]
@zanderso do you think this is safe to merge?

**Comment by zanderso** [comment]
Yes. The flag is off by default, so I suspect no device lab tests need to be updated as part of this PR.

**Pull request merged**